### PR TITLE
Fix network merge-detach-merge issue

### PR DESCRIPTION
**Pull request by jeandemanged**
**Please check if the PR fulfills these requirements**

- [X] The commit message follows our guidelines
- [X] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)
- [ ] A PR or issue has been opened in all impacted repositories (if any)

**Does this PR already have an issue describing the problem?**

No

**What kind of change does this PR introduce?**

Bug fix

**What is the current behavior?**

When some networks are merged, then detached and re-merged, their substations are not detected as part of the their respective sub-networks.

**What is the new behavior (if this is a feature change)?**
In the merge - detach - merge use case, the substations are now correctly detected as elements of their respective sub-networks.

**Does this PR introduce a breaking change or deprecate an API?**
- [ ] Yes
- [X] No

**If yes, please check if the following requirements are fulfilled**

- [ ] The *Breaking Change* or *Deprecated* label has been added
- [ ] The migration steps are described in the following section

**What changes might users need to make in their application due to this PR? (migration steps)**

**Other information**:

The problem was located in the "detach" operation. After this operation, the `subnetworkRef` attribute of the substations was not pointing to the same `RefObj` object as the `subnetworkRef` attribute of the detached network.
This was problematic when the detached network was then merged because the object referenced by the `subnetworkRef` attribute of the substations was then not "automatically updated", as expected.

### CODE REVIEW

File: iidm/iidm-impl/src/test/java/com/powsybl/iidm/network/impl/MergeTest.java [hunk 1]
```diff
@@ -199,6 +199,58 @@ void mergeNetworksWithSameCaseDates() {
...
+        Network net1c = merged2.getSubnetwork("N1");
+        Network net2c = merged2.getSubnetwork("N2");
+        assertEquals(merged2, sub1.getNetwork());
+        assertEquals(net1c, sub1.getParentNetwork());
+        assertEquals(merged2, sub2.getNetwork());
+        assertEquals(net2c, sub2.getParentNetwork());
+        assertEquals(2, merged2.getSubstationCount());
+        assertEquals(1, net1c.getSubstationCount());
+        assertEquals(1, net2c.getSubstationCount());
+    }
```
Comment (rolnico) [inline iidm/iidm-impl/src/test/java/com/powsybl/iidm/network/impl/MergeTest.java:246]: I would extract the assertions in two separate methods:

```suggestion
    void mergeDetachMergeTest() {
        // two networks
        Network net1 = NetworkFactory.findDefault().createNetwork("N1", "test");
        Substation sub1 = net1.newSubstation().setId("S1").add();
        assertIndependentNetwork(net1, sub1);

        Network net2 = NetworkFactory.findDefault().createNetwork("N2", "test");
        Substation sub2 = net2.newSubstation().setId("S2").add();
        assertIndependentNetwork(net2, sub2);

        // merge them
        Network merged1 = Network.merge(net1, net2);
        Network net1a = merged1.getSubnetwork("N1");
        Network net2a = merged1.getSubnetwork("N2");
        assertMergedNetworks(merged1, net1a, sub1, net2a, sub2);

        // detach them
        Network net1b = net1a.detach();
        assertIndependentNetwork(net1b, sub1);
        assertEquals(1, merged1.getSubstationCount());
        Network net2b = net2a.detach();
        assertIndependentNetwork(net2b, sub2);
        assertEquals(0, merged1.getSubstationCount());

        // merge again
        Network merged2 = Network.merge(net1b, net2b);
        Network net1c = merged2.getSubnetwork("N1");
        Network net2c = merged2.getSubnetwork("N2");
        assertMergedNetworks(merged2, net1c, sub1, net2c, sub2);
    }

    private static void assertIndependentNetwork(Network network, Substation substation) {
        assertEquals(network, substation.getNetwork());
        assertEquals(network, substation.getParentNetwork());
        assertEquals(1, network.getSubstationCount());
    }

    private static void assertMergedNetworks(Network mergedNetwork,
                                             Network network1, Substation substation1,
                                             Network network2, Substation substation2) {
        assertEquals(mergedNetwork, substation1.getNetwork());
        assertEquals(network1, substation1.getParentNetwork());
        assertEquals(mergedNetwork, substation2.getNetwork());
        assertEquals(network2, substation2.getParentNetwork());
        assertEquals(2, mergedNetwork.getSubstationCount());
        assertEquals(1, network1.getSubstationCount());
        assertEquals(1, network2.getSubstationCount());
    }
```

Comment (jeandemanged) [reply iidm/iidm-impl/src/test/java/com/powsybl/iidm/network/impl/MergeTest.java:246]: done, thank you

**Review by rolnico — APPROVED** [review]
